components/browser/toolbar/src/main/java/mozilla/components/browser/toolbar/BrowserToolbar.kt

```diff
@@ -163,17 +190,51 @@ class BrowserToolbar @JvmOverloads constructor(
         // urlView decides when it's appropriate to ask for autocompletion, and in turn we invoke
         // our 'filter' and send results back to 'urlView'.
         editToolbar.urlView.setOnFilterListener { text ->
-            filter.invoke(text, object : AutocompleteDelegate {
-                override fun applyAutocompleteResult(result: AutocompleteResult) {
-                    editToolbar.urlView.applyAutocompleteResult(InlineAutocompleteEditText.AutocompleteResult(
-                        text = result.text, source = result.source, totalItems = result.totalItems
-                    ))
-                }
-
-                override fun noAutocompleteResult() {
-                    editToolbar.urlView.noAutocompleteResult()
-                }
-            })
+            // Keep track of the latest query to throw away stale autocompletion results.
+            latestAutocompleteInput = text
+
+            // We got a new input, so whatever past autocomplete queries we still have running are
+            // irrelevant. We cancel them, but do not depend on cancellation to take place.
+            autocompleteSupervisorJob.cancelChildren()
+
+            // Our various autocomplete providers may perform expensive tasks to field our queries,
+            // such as disk access. Let's do the actual autocompletion work off the UI thread.
+            CoroutineScope(autocompleteDispatcher).launch {
+                filter.invoke(text, object : AutocompleteDelegate {
+                    override fun applyAutocompleteResult(result: AutocompleteResult) {
+                        // Ignore this result if the query is stale.
+                        if (result.input != latestAutocompleteInput) {
+                            Log.log(
+                                tag = logTag,
+                                priority = Log.Priority.INFO,
+                                message = "Discarding result for \"${result.input}\" input"
+                            )
+                            return
+                        }
+                        // Jump back to the UI thread to process the results.
+                        CoroutineScope(Dispatchers.Main).launch {
```

Comment:
You are not actually jumping back to the UI thread but launching a new asynchronous coroutine here; of which there can be multiple. Maybe you want to use [withContext](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/with-context.html) here?